torchdynamo/_eval_frame.c

```diff
@@ -216,12 +228,36 @@ static PyCodeObject *lookup(CacheEntry *e, PyObject *f_locals) {
   }
   Py_DECREF(valid);
   if (valid == Py_True) {
-    return e->code;
+    CacheEntry *extra = get_extra(frame->f_code), *next = e->next, *head = extra;
+    if (extra != e) {
+      if (next) {
+        PyObject *check_fn_tmp = e->check_fn;
+        PyCodeObject *code_tmp = e->code;
+        memcpy(e, next, sizeof(CacheEntry));
+        next->check_fn = check_fn_tmp;
+        next->code = code_tmp;
+        next->next = extra;
+        head = next;
+      } else {
+        CacheEntry *cur = extra;
+        while (cur) {
```

Comment:
We should be able to do this in O(1) time.  We would need a pred pointer for that though.

It would also be a lot cleaner if we could handle it with a single case rather than two.

Reply:
Hi, thanks for your advice! As you know, the caches are managed by a single linked list, so it is difficult to adjust the data structure directly. If possible, we can change it into a double linked list for O(1) and less case handling.

Reply:
One alternative to making it double-linked would be to pass the pred pointer of the current node in as an arg to the lookup function.  Since we are walking the list, we should already know the prior node (it was the one we looked at right before the current one).